core/src/main/java/io/parsingdata/metal/token/CycleToken.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2013-2020 Netherlands Forensic Institute
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.parsingdata.metal.token;
+
+import java.util.Optional;
+
+import io.parsingdata.metal.data.Environment;
+import io.parsingdata.metal.data.ParseReference;
+import io.parsingdata.metal.data.ParseState;
+import io.parsingdata.metal.encoding.Encoding;
+
+/**
+ * Base class for {@link Token} implementations that can potentially
+ * cause a cycle if used in conjunction with {@link Sub}.
```

Comment:
What does `in conjunction with Sub` mean?

I understand the change of the PR, but this Javadoc seems a bit vague.

Maybe we can elaborate more what the `CycleToken`s typically are, or how they behave, maybe something like "tokens that contain other tokens" or "tokens that use other tokens as building blocks" (also pretty vague, I admit).

Reply:
And related to that: Which tokens are *not* `CycleToken`s?

Reply:
Before you know it, we're back in the discussion around whether it's actually a non-terminal or not. But I agree it should be explained to the user here!

Reply:
I've reworded the description and expanded it to explicitly explain that all non-terminal `Token`s are of this type except `TokenRef` and `Tie` along with the rationale for that.